### PR TITLE
Load plugins with require/import

### PR DESCRIPTION
**Pull request by alcuadrado**
This plugins updates Hardhat core and the plugins so that plugins are now loaded using `require`/`import`.

There are a few things missing, but I decided to include them in another PR so it's easier to review them. They are:
* Updating the READMEs
* Updating the plugins to use a single tsconfig.json, as that's less error prone, and was needed in `hardhat-truffle5`.
* Updating the plugins' tests to clearly distinguish which are test files and which are fixture projects, as randomly loading a fixture project's config file results in an error.
* Throwing an error when hardhat's cli is used and not installed.
* Update the project creation module.

### CODE REVIEW

**Comment by alcuadrado** [comment]
I wish I could promise this is going to be the last PR this big :(

File: packages/hardhat-truffle5/src/index.ts [hunk 1]
```diff
@@ -22,170 +23,187 @@ import { LazyTruffleContractProvisioner } from "./provisioner";
...
-      const web3Path = require.resolve("web3");
-      const formattersPath = require.resolve(
-        "web3-core-helpers/src/formatters",
-        {
-          paths: [web3Path],
+    );
+
+    monkeyPatchMethod(
+      execute,
+      "originalGetGasEstimate",
```
Comment (fvictorio) [inline packages/hardhat-truffle5/src/index.ts:90]:
```suggestion
      "getGasEstimate",
```

File: packages/hardhat-core/src/internal/core/execution-mode.ts [hunk 2]
```diff
@@ -1,89 +1,23 @@
...
-    if (__dirname.startsWith(workingDirectoryOnLoad)) {
-      return ExecutionMode.EXECUTION_MODE_TS_NODE_TESTS;
-    }
-
-    return ExecutionMode.EXECUTION_MODE_LINKED;
-  }
-
+export function isHardhatInstalledLocally(configPath?: string) {
   try {
-    if (require("is-installed-globally")) {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/core/execution-mode.ts:36]: Yes 🥳

File: packages/hardhat-core/src/internal/core/config/config-loading.ts [hunk 3]
```diff
@@ -1,16 +1,14 @@
 import chalk from "chalk";
-import findUp from "find-up";
 import fsExtra from "fs-extra";
 import path from "path";
 import * as stackTraceParser from "stacktrace-parser";
 
 import { HardhatArguments, HardhatConfig } from "../../../types";
 import { HardhatContext } from "../../context";
 import { findClosestPackageJson } from "../../util/packageInfo";
+import { getRequireCachedFiles } from "../../util/platform";
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/core/config/config-loading.ts:9]: I wish the linter detected this :(

File: packages/hardhat-web3/src/type-extensions.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import "hardhat/types/runtime";
+import type Web3 from "web3";
 
 declare module "hardhat/types/runtime" {
   interface HardhatRuntimeEnvironment {
-    Web3: any;
-    web3: any;
+    Web3: typeof Web3;
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-web3/src/type-extensions.ts:6]: It took me some time to understand this originally. `Web3` is a class, so both a type and a value. If you do `const w : Web3`, it's used a a type, so `w`'s type if that of an instance of `Web3`. If you want the class' type (i.e. the type of the class object), you need to do a `typeof Web3`.

File: packages/hardhat-core/test/internal/core/config/config-loading.ts [hunk 5]
```diff
@@ -309,4 +309,49 @@ Hardhat plugin instead.`
...
+        assert.include(files, path.join(projectPath, "b.js"));
+
+        // Must not include unrelated files.
+        assert.notInclude(files, path.join(projectPath, "not-imported.js"));
+
+        resetHardhatContext();
+      }
+    });
+
+    afterEach(function () {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/test/internal/core/config/config-loading.ts:353]: Agreed!

**Pull request merged**